src/Text/Layout/Table/Cell/MultiFormatted.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE DeriveTraversable #-}
+
+-- | Provides formatting to an instance of 'Cell'. For example, in a unix
+-- terminal one could use the following:
+--
+-- >>> buildCell (formatted "\ESC[31m" "Hello World!" "\ESC[0m") :: String
+-- Hello World!
+--
+-- The text then appears in dull red.
+module Text.Layout.Table.Cell.MultiFormatted
+    ( MultiFormatted
+    , multiFormatted
+    , formatted
+    , plain
+    ) where
+
+import Data.List (foldl', mapAccumL, mapAccumR)
+import Data.String
+
+import Text.Layout.Table.Primitives.AlignInfo
+import Text.Layout.Table.Cell
+import Text.Layout.Table.StringBuilder
+import Text.Layout.Table.StringBuilder
+
+data MultiFormatted a
+    = MultiFormatted
+      { prefix      :: String
+      , contentList :: [MultiFormatted a]
+      , suffix      :: String
+      }
+    | PlainContent a
```

Comment:
Do you think it would bring any improvement if we represented it as:
```
data MultiFormatted a
    = Plain a
    | Concat [a]
    | Empty
    | Format String (MultiFormatted a) String
```
Then one could just provide `plain`, `formattedMF :: String -> MultiFormatted a -> String -> MultiFormatted a`, and the instances.  It feels like an easier interface for the user.
And we could provide an associative `<>`:
```
Empty <> b           =
a           <> Empty = b
a           <> b     = Concat $ elements a <> elements b
  where
    elements (Concat es) = es
    elements e = [e]
```

Reply:
We could do. It would mean that there's some redundancy in the data constructors, as `Format "" a "" == a` after projecting through the Cell class methods, but I agree that your representation seems more natural. I'll make the change unless you change your mind.

Reply:
Also, there's a redundancy between `Empty` and `Concat []` and between `Plain a` and `Concat . pure`.

Reply:
I've mostly adopted your suggestions, but I've left out the Empty constructor. It's role is completely fulfilled by `Concat []`.

Reply:
> It would mean that there's some redundancy in the data constructors, as Format "" a "" == a after projecting through the Cell class methods,

Can you give an example where that would happen?

> I've mostly adopted your suggestions, but I've left out the Empty constructor. It's role is completely fulfilled by Concat [].

I guess that works.  As an alternative we may employ `Data.List.NonEmpty`. :)

Reply:
> > It would mean that there's some redundancy in the data constructors, as Format "" a "" == a after projecting through the Cell class methods,
> 
> Can you give an example where that would happen?

It wouldn't unless somebody were doing something silly. I think we can safely ignore this.

> > I've mostly adopted your suggestions, but I've left out the Empty constructor. It's role is completely fulfilled by Concat [].
> 
> I guess that works. As an alternative we may employ `Data.List.NonEmpty`. :)

That is another option, but I don't really want to introduce another data constructor whose explicit purpose is not to be used directly. If we're not exposing the constructor anyway, I don't think we need to introduce it.

But it's your call. Let me know and I'll implement it.

Reply:
* I think it is elegant to have the neutral element as an extra constructor.
* Matching on the empty list will have to follow at least one pointer, whereas an extra constructor can be represented within the pointer.
* Having something like `Concat []` and `Concat [foo]` is weird because there is nothing really concatenated.  This is especially true as long as we have a `Show` instance.

I think the first two do not really matter that much.

Reply:
If you feel strongly about it I'm happy to add in `Empty` again.

This seems to be implying that we should restrict `Concat` to having at least two elements. It's constructor would then be `Concat a a [a]`. This definitely obfuscates the purpose of the constructor.

Reply:
For comparison, the way doclayout handles this is with an `Empty` constructor, and a `Concat` constructor which takes exactly two arguments. This is elegant, but again their `Semigroup` instance is technically not lawful, as it is not associative. They seem to get by just fine ignoring this technicality.

Reply:
I've added `Empty` back in according to your original specification.